server/analysis/tests/simulation.py

```diff
@@ -0,0 +1,145 @@
+        return reward, metric_data
+
+    def simulate(self, knob_data):
+        if self.mode == 0:
+            return self.simulate_identity_sqrt(knob_data)
+        elif self.mode == 1:
+            return self.simulate_threshold(knob_data)
+        elif self.mode == 2:
+            return self.borehole(knob_data)
+
+
+def train_ddpg(env, gamma=0.99, tau=0.002, lr=0.01, batch_size=32, n_loops=1000):
+    results = []
+    x_axis = []
+    ddpg = DDPG(n_actions=env.knob_dim, n_states=env.metric_dim, gamma=gamma, tau=tau,
+                clr=lr, alr=lr, batch_size=batch_size)
+    knob_data = np.random.rand(env.knob_dim)
+    prev_metric_data = np.zeros(env.metric_dim)
+    for i in range(n_loops):
+        reward, metric_data = env.simulate(knob_data)
+        ddpg.add_sample(prev_metric_data, knob_data, reward, metric_data, False)
+        ddpg.update()
+        if i % 20 == 0:
+            results.append(test_ddpg_ypreds(env, ddpg))
+            x_axis.append(i)
+        prev_metric_data = metric_data
+        knob_data = ddpg.choose_action(prev_metric_data)
+    return np.array(results), np.array(x_axis)
+
+
+def test_ddpg_ypreds(env, ddpg):
+    total_reward = 0.0
+    n_samples = 100
+    prev_metric_data = np.zeros(env.metric_dim)
+    for _ in range(n_samples):
+        knob_data = ddpg.choose_action(prev_metric_data)
+        reward, prev_metric_data = env.simulate(knob_data)
+        total_reward += reward
+    return total_reward / n_samples
+
+
+def plotline(x_axis, data, title, path):
+    plt.plot(x_axis, data)
+    plt.xlabel("loops")
+    plt.ylabel("rewards")
+    plt.title(title)
+    plt.savefig(path)
+    plt.clf()
+
+
+def plotlines(x_axis, data1, data2, label1, label2, title, path):
+    plt.plot(x_axis, data1, color='red', label=label1)
+    plt.plot(x_axis, data2, color='blue', label=label2)
+    plt.legend()
+    plt.xlabel("loops")
+    plt.ylabel("rewards")
+    plt.title(title)
+    plt.savefig(path)
+    plt.clf()
+
+
+def main(knob_dim=192, lr=0.001):
+    random.seed(0)
+    np.random.seed(1)
+    torch.manual_seed(1)
+    metric_dim = 60
+    env = Environment(knob_dim, metric_dim, mode=0)
+    n_loops = 3000
+    if not os.path.exists("figures"):
+        os.mkdir("figures")
+
+    n_repeats = 10
+    for i in range(n_repeats):
+        if i == 0:
+            results1, x_axis = train_ddpg(env, gamma=0, lr=lr, n_loops=n_loops)
+        else:
+            results1 += train_ddpg(env, gamma=0, lr=lr, n_loops=n_loops)[0]
+    for i in range(n_repeats):
+        if i == 0:
+            results2, x_axis = train_ddpg(env, gamma=0.99, lr=lr, n_loops=n_loops)
+        else:
+            results2 += train_ddpg(env, gamma=0.99, lr=lr, n_loops=n_loops)[0]
+    results1 /= n_repeats
+    results2 /= n_repeats
+    title = "knob_dim:{}_lr:{}".format(knob_dim, lr)
+    filename = "figures/{}.pdf".format(title)
+    plotlines(x_axis, results1, results2, "gamma=0", "gamma=0.99", title, filename)
+
```

Comment:
You could additionally save the data needed to plot the results in json/csv/npz/etc format if matplotlib isn't installed (or always save it)